server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1160,23 +1160,28 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         return snapshotEntry;
     }
 
-    /**
-     * Deletes snapshots that are assumed to be in the repository and not tracked as in-progress in the cluster state.
-     *
-     * @param snapshotIds       Snapshots to delete
-     * @param repoName          Repository name
-     * @param repositoryStateId Repository generation to base the delete on
-     * @param listener          Listener to complete when done
-     */
-    private void deleteCompletedSnapshots(List<SnapshotId> snapshotIds, String repoName, long repositoryStateId, Priority priority,
-                                          ActionListener<Void> listener) {
+    private ClusterStateUpdateTask createDeleteStateUpdate(List<SnapshotId> snapshotIds, String repoName, long repositoryStateId,
+                                                           Priority priority, ActionListener<Void> listener) {
+        // Short circuit to noop state update if there isn't anything to delete
         if (snapshotIds.isEmpty()) {
-            listener.onResponse(null);
-            return;
+            return new ClusterStateUpdateTask() {
```

Comment:
This is a little less ergonomic than I would've liked but I found it easiest to produce a CS update task in the `Repository` API compared to other mechanism because the existing code is so heavily built on using state in CS update tasks to do logic between `execute` and `clusterStateProcessed`, so I figured doing it this way and in the repository allows for the follow-ups to be relatively isolated changes.